### PR TITLE
Make adcc fully open-source

### LINKED ISSUE #105: Move adccore to adcc
Just an issue to keep track of things that need to be done:

- [x] merge #71 
- [x] merge #104 
- [x] Deploy libtensor to github release (for MacOS and Linux)
- [x] Deploy libtensor to conda (WIP already)
- [x] Merge codes
- [x] Restructure repo
- [x] Update deployment infrastructure
- [x] transfer issues from core repository

### PR DESCRIPTION
**Pull request by mfherbst**
PR to track the progress of making adcc fully open-source. Closes #105.

## Progress
- [x] Simplify the number of define flags and commandline flags needed to talk to libtensor
- [x] Get libtensor from github releases
- [x] MacOS support
- [x] C++ tests build
- [x] C++ tests integrated into setup.py test
- [x] Better file structure (libadcc/adcc is a bit weird)
- [x] Timings (to check we have not messed up)
  - [x] "normal" build
  - [x] conda build
- [x] Update setup.py installation (manifest, installed files, package files, ...)
- [x] Update scripts (grep for adccore)
- [x] Fix building docs
- [x] Update documentation
- [x] Conda support and upload
- [x] Merge deployment jobs into single file & enable all py-versions for Conda
- ~~[ ] LGTM for C++~~ ➡️ postpone
- [x] Coverage for C++
- [x] Code-style test for C++ using clang-format

### CODE REVIEW

File: run_tests.sh
Comment (maxscheurer) [inline run_tests.sh:33]: Are these paths still correct?

Comment (mfherbst) [reply run_tests.sh:33]: Obviously not. The only thing that works right now is setup.py test.

Comment (maxscheurer) [reply run_tests.sh:33]: Yes, just wanted to add a note that we don't forget...

**Comment by maxscheurer** [comment]
In `export_Tensor.cc` [here](https://github.com/adc-connect/adcc/blob/733d32442a476167d42cb22d7389c4a9bf31f561/libadcc/export_Tensor.cc#L38)
can you do
```C++
for (auto tpl : permutations) {
    if (py::isinstance<py::int_>(tpl)) {
      iterator_of_ints = true;
      break;
    }
  }
```

File: setup.py
Comment (maxscheurer) [inline setup.py:303]: Classic.... 🙃

Comment (mfherbst) [reply setup.py:303]: Indeed, some tricks you know after a while. Let's hope that does it.

**Comment by maxscheurer** [comment]
Can I make a commit here or do you have WIP?

**Comment by mfherbst** [comment]
I won't work on it tonight and the next things I have are to get the C++ tests working. So that's mostly stuff in libadcc and setup.py.

File: setup.py
Comment (mfherbst) [inline setup.py:250]: Are you sure this is the "right" thing to test for. Isn't this rather a clang-specific thing?

File: .github/workflows/deploy_conda.yaml
Comment (mfherbst) [inline .github/workflows/deploy_conda.yaml:1]: I think it would be a good idea to merge this with the deploy workflow because they are essentially triggered under the same circumstances.

File: setup.py
Comment (maxscheurer) [reply setup.py:250]: Kind of... AppleClang does not complain, and the only macOS compiler on `conda` is clang... I can add a note though.

File: .github/workflows/deploy_conda.yaml
Comment (maxscheurer) [reply .github/workflows/deploy_conda.yaml:1]: Yes, I think GHA can still run them as separate "jobs", right?

Comment (mfherbst) [reply .github/workflows/deploy_conda.yaml:1]: sure. You can even make the jobs depend on each other. That might allow us to only run the tests once.

**Comment by maxscheurer** [comment]
We also should disable automatic download of libtensorlight in conda builds, such that we don't accidentally link against our "self-provided" version instead of the conda one.

File: setup.py
Comment (mfherbst) [reply setup.py:250]: But what if I use clang on ubuntu?

**Comment by mfherbst** [comment]
You can. Just provide a custom siteconfig.

**Comment by maxscheurer** [comment]
But then I have to create an extra file just for the `conda` deployment?

**Comment by mfherbst** [comment]
Sure. But I would write it to the home, that's more reliable (you don't depend on the working dir). It searches for `~/.adcc/siteconfig.py` or so ... check the setup.py.

File: setup.py
Comment (maxscheurer) [reply setup.py:250]: Yes... so we need to get the compiler name and check for that 😄

Comment (maxscheurer) [reply setup.py:250]: Mhm do you know by heart how to get the compiler name outside the extension class? 🤕

Comment (maxscheurer) [reply setup.py:250]: Otherwise, I think one can just add the two flags directly to `extra_compile_args` without a compiler/platform distinction.

Comment (mfherbst) [reply setup.py:250]: > how to get the compiler name outside the extension class?

No, but I guess you can always get it *after* the extension has been created (then the user can't override it, but at least things compile.

I think in order to decide what to do you will need to test if this problem also occurs in other circumstances. If it's conda-specific I don't mind the way you solved it now. If it's not then I think it should be solved more generally.

(On that note I think the siteconfig solution is not the best way to tackle it, better disable it in the setup.py)

Comment (maxscheurer) [reply setup.py:250]: It seems to be conda+darwin-specific. Clang on arch compiles without complaining.

Comment (mfherbst) [reply setup.py:250]: Ok cool. Then I'm happy as it is.

Comment (maxscheurer) [reply setup.py:250]: Giving conda another run with the current configuration... Then I can also run some timings using the conda builds 😄

**Comment by maxscheurer** [comment]
If you make commits here, can you leave the conda builds turned on for the time being? I need a fresh adcc on dev (because of the breaking lt allocator changes).

File: scripts/upload_to_anaconda.sh
Comment (maxscheurer) [inline scripts/upload_to_anaconda.sh:37]: Can now be changed to `-c adcc`

**Comment by maxscheurer** [comment]
Nice twist, conda, not wanting the channel name that is accidentally identical with a local directory 😂😂😂

**Comment by mfherbst** [comment]
@maxscheurer Do you know how to exclude a file from LGTM. The siteconfig_example.py should be excluded from the analysis, I'd say.

**Comment by maxscheurer** [comment]
I think configuring LGTM is really tough, but maybe it works via a [lgtm.yml](https://help.semmle.com/lgtm-enterprise/user/help/lgtm.yml-configuration-file.html) config file?

**Comment by mfherbst** [comment]
> I think configuring LGTM is really tough, but maybe it works via a lgtm.yml config file?

Could be, I can't be asked to crawl through the docs at the moment though. Would you be willing to take care of that. I tried the C++ thing a few days back, but it did not work out of the box so I have not yet done it either.

**Comment by mfherbst** [comment]
Ah ... ok that's surpirisingly easy.

**Comment by maxscheurer** [comment]
> Ah ... ok that's surpirisingly easy.

I don't know if it works...

BTW did you test the `siteconfig.py` by specifying `libtensor_autoinstall` to some other path than the default? On my workstation the file is read, but the `libtensor_autoinstall` I set is being ignored.

**Comment by mfherbst** [comment]
I know of that problem. I tested it, but before a refactor that broke it. It's now fixed.

**Comment by maxscheurer** [comment]
> I know of that problem. I tested it, but before a refactor that broke it. It's now fixed in the C++ coverage commit.

I fixed a typo in the setup, so it's working now 👍

**Comment by maxscheurer** [comment]
I bet LGTM for C++ will only work once this PR is merged into master, otherwise, I think there's no way to trigger another language on a PR...

**Comment by mfherbst** [comment]
I guess we do it later than.

**Comment by maxscheurer** [comment]
It seems like LGTM is just ignoring the `.lgtm.yml` file 😒
Hopefully it works when the PR is merged, but currently this feels like a waste of time.

**Comment by mfherbst** [comment]
Well now that I record C++ coverage the python coverage is ignored ^^.

**Comment by maxscheurer** [comment]
Question: Should we disable running the full test suite in conda builds? Instead, we could only run them during pip deployment and make the conda jobs depend on successful pip deployment...

Another suggestion is to just run tests on a single case in conda (e.g. `h2o_sto3g`) to make sure the built `adcc` in principle works.

If we support 4 Python versions (3.6-3.9, like Psi4) and 2 OS, we'd run the entire tests 8x during the deployment...

**Comment by mfherbst** [comment]
> Another suggestion is to just run tests on a single case in conda (e.g. h2o_sto3g) to make sure the built adcc in principle works.

That plus the dependency on the pip job sounds like a good compromise. Also I suppose on the master / dev conda packages we can be more restrictive and e.g. exclude MacOS, which takes forever and perhaps some older python versions.

**Comment by maxscheurer** [comment]
> we can be more restrictive and e.g. exclude MacOS, which takes forever and perhaps some older python versions.

You mean wrt the test runs?

**Comment by mfherbst** [comment]
No with the actual dev deployments.

**Comment by mfherbst** [comment]
How are the timings by the way?

**Comment by maxscheurer** [comment]
> No with the actual dev deployments.

Okay, I don't have a good idea how to hack this in right now...

**Comment by maxscheurer** [comment]
> How are the timings by the way?

Looking good, I'm running another large job right now, but I'd say we did not loose any performance on the way 🚀

**Comment by maxscheurer** [comment]
Okay, I've adapted the workflow for conda to only deploy py38 to `dev`. I'd like to give this maybe another try and see if it works..

**Comment by mfherbst** [comment]
Yeah sure. I'm not going to work on this today any more. Also feel free to take a round at cleaning up the docs if you want :smile:.

**Comment by maxscheurer** [comment]
Okay, why not 😄 How should the "adccore" stuff be called from now on?

**Comment by mfherbst** [comment]
Basically `libadcc`, but quite some of it will just need to go (e.g. overview of the structure of the code talks about the ADC equations being implemented in C++ and all that).

Like in the API documentation we distinguish libadcc and adccore and now libadcc has a c++ and a python part, so a bit of restructuring is needed there. Just come up with a sensible solution. The docs are pretty messy at the moment anyways. We really really should have a solid pass where we think about a more sensible structure. Probably less should be in static html and more in interactive examples anyway, but that I suppose is for another day.

**Comment by maxscheurer** [comment]
I've made a pass at the docs, such that no more "wrong information" is present in the documentation 😄

**Comment by maxscheurer** [comment]
Conda packaging on Linux now takes 12m, 22m on macOS 🍎 ... that's nice 🚀 🎸

**Review by maxscheurer — APPROVED** [review]

**Comment by mfherbst** [comment]
Cool. I'll try to make anther pass at the docs today or tomorrow and than this can go in from my end as well.

**Review by maxscheurer — APPROVED** [review]
Huge effort over a long time that led to this PR 🥳
Thanks for the joint effort!

**Comment by mfherbst** [comment]
Yes pretty much half a year, no? I'm really happy we're close to merging. Thanks for the teamwork! :fireworks:

Too bad the situation and all that but we should have some :beers: some time over this

**Comment by maxscheurer** [comment]
I'm having a 🍺 now already, but yes! Hopefully adcc-🍻 soon...

**Comment by mfherbst** [comment]
:smile:, I'll get one, too ;).

**Comment by maxscheurer** [comment]
Can I bump to `v0.15.6`? 😄

**Comment by maxscheurer** [comment]
Or you can do it as well, I'm upgrading macOS now... 🔥

**Comment by mfherbst** [comment]
Oh yikes ... have fun! Would be good to see if our code then still works ;)

By the way: There is actually an issue with our coda deployment, because in the current workflow version 0.15.6 would be deployed both as 0.15.6.dev (to the dev label) and  0.15.6 (to the main label), which does not make super much sense. I'll try to manually cancel that this time, but that we should fix.

**Comment by maxscheurer** [comment]
Okay, if you cannot cancel it, I can remove the files from dev label. I don't know how to fix it yet...

**Comment by mfherbst** [comment]
No worries I cancelled, but yes that's a bit annoying. You can mark a complete job with an `if:` in GHA. I suppose that's what we have to do somehow (like if it's building on the master branch and got a tag, than don't run coda deployment only if it's building on the tag branch and got a tag). But don't ask me how that would work.

**Pull request merged**
